core/shared/src/main/scala/zio/ZLayer.scala

```diff
@@ -413,57 +410,76 @@ object ZLayer {
-    ): ZManaged[A, E, B] =
-      ZManaged {
-        get(layer).flatMap {
-          case Some(b) => ZIO.succeedNow(Reservation(ZIO.succeedNow(b), _ => UIO.unit))
-          case None =>
-            layer.scope.flatMap(_.apply(self)).reserve.map { reservation =>
-              Reservation(
-                reservation.acquire.tap(b => memoize(layer, b)),
-                _ =>
-                  ZIO.accessM[A] { a =>
-                    finalizerRef.update { finalizer => exit =>
-                      finalizer(exit) *> reservation.release(exit).provide(a)
-                    }
-                  }
-              )
-            }
-        }
-      }
+    ): ZManaged[A, E, B]
   }
 
   private object MemoMap {
 
     /**
-     * Constructs an empty memo map backed by a `Ref`.
+     * Constructs an empty memo map.
      */
     def make: UIO[MemoMap] =
-      Ref.make[Map[ZLayer[Nothing, Any, Has[_]], Any]](Map.empty).map { ref =>
-        new MemoMap {
-          def get[E, A, B <: Has[_]](layer: ZLayer[A, E, B]): UIO[Option[B]] =
-            ref.get.map(_.get(layer).asInstanceOf[Option[B]])
-          def memoize[E, A, B <: Has[_]](layer: ZLayer[A, E, B], b: B): UIO[Unit] =
-            ref.update(_ + (layer -> b))
+      RefM
+        .make[Map[ZLayer[Nothing, Any, Has[_]], (Promise[_, _], Exit[Any, Any] => ZIO[Any, Nothing, Any], Ref[Int])]](
+          Map.empty
+        )
+        .map { ref =>
+          new MemoMap { self =>
+            final def getOrElseMemoize[E, A, B <: Has[_]](
+              layer: ZLayer[A, E, B],
+              finalizerRef: Ref[Exit[Any, Any] => ZIO[Any, Nothing, Any]]
+            ): ZManaged[A, E, B] =
+              ZManaged {
+                ref.modify { map =>
+                  map.get(layer) match {
+                    case Some((promise, release, observers)) =>
+                      ZIO.succeedNow {
+                        (
+                          Reservation(
+                            observers.update(_ + 1) *>
+                              promise.await.bimap(_.asInstanceOf[E], _.asInstanceOf[B]),
+                            _ =>
+                              finalizerRef.update { finalizer => exit =>
+                                finalizer(exit) *>
+                                  release(exit).whenM(observers.updateAndGet(_ - 1).map(_ == 0))
+                              }
+                          ),
+                          map
+                        )
+                      }
+                    case None =>
+                      for {
+                        a           <- ZIO.environment[A]
+                        promise     <- Promise.make[E, B]
+                        observers   <- Ref.make(0)
+                        reservation <- layer.scope.flatMap(_.apply(self)).reserve
+                      } yield (
+                        Reservation(
+                          observers.update(_ + 1) *>
```

Comment:
I'm wondering if you should maybe use `uninterruptibleMask` to make sure the `observers` ref is always updated?

Reply:
I think we are okay because `observers` is always updated in either the `acquire` or `release` action of a `Managed`.